security/security-spring/src/main/java/org/cbioportal/security/spring/CancerStudyPermissionEvaluator.java

```diff
@@ -262,6 +315,119 @@ public boolean hasPermission(Authentication authentication, Object targetDomainO
+            studyIds.add(identifier.getStudyId());
+        }
+        return hasAccessToCancerStudies(authentication, studyIds, permission);
+    }
+
+    private boolean hasAccessToCancerStudies(Authentication authentication, PatientFilter patientFilter, Object permission)
+    {
+        // use hashset as this list can be populated with many duplicate values
+        Set<String> studyIds = new HashSet<String>();
+        if (patientFilter.getPatientIdentifiers() != null) {
+            for (PatientIdentifier patientIdentifier : patientFilter.getPatientIdentifiers()) {
+                studyIds.add(patientIdentifier.getStudyId());
+            }
+        }
+        else {
+            uniqueKeyExtractor.extractUniqueKeys(patientFilter.getUniquePatientKeys(), studyIds);
+        }
+        return hasAccessToCancerStudies(authentication, studyIds, permission);
+    }
+
+    private boolean hasAccessToCancerStudies(Authentication authentication, SampleFilter sampleFilter, Object permission)
+    {
+        if (sampleFilter.getSampleListIds() != null) {
+            return hasAccessToSampleLists(authentication, sampleFilter.getSampleListIds(), permission);
+        }
+        else if (sampleFilter.getSampleIdentifiers() != null) {
+            return hasAccessToCancerStudiesBySampleIdentifier(authentication, sampleFilter.getSampleIdentifiers(), permission);
+        }
+        else {
+            // use hashset as this list can be populated with many duplicate values
+            Set<String> studyIds = new HashSet<String>();
+            uniqueKeyExtractor.extractUniqueKeys(sampleFilter.getUniqueSampleKeys(), studyIds);
+            return hasAccessToCancerStudies(authentication, studyIds, permission);
+        }
+    }
+
+    private boolean hasAccessToCancerStudiesBySampleIdentifier(Authentication authentication, List<SampleIdentifier> sampleIdentifiers, Object permission)
+    {
+        // use hashset as this list can be populated with many duplicate values
+        Set<String> studyIds = new HashSet<String>();
+        for (SampleIdentifier sampleIdentifier : sampleIdentifiers) {
+            studyIds.add(sampleIdentifier.getStudyId());
+        }
+        return hasAccessToCancerStudies(authentication, studyIds, permission);
+    }
+
+    private boolean hasAccessToCancerStudies(Authentication authentication, Collection<String> studyIds, Object permission)
+    {
+        for (String studyId : studyIds) {
+            if (!hasPermission(authentication, studyId, "CancerStudy", permission)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean hasAccessToMolecularProfiles(Authentication authentication, Collection<String> molecularProfileIds, Object permission)
+    {
+        for (String molecularProfileId : molecularProfileIds) {
+            MolecularProfile molecularProfile = molecularProfileRepository.getMolecularProfile(molecularProfileId);
```

Comment:
we can use fetch methods to get multiple profiles in one call, to prevent making db calls in a loop. (the optimization I mentioned a few weeks ago)

Reply:
Yes, good idea. I can do this.